agent/structs/config_entry_discoverychain.go

```diff
@@ -0,0 +1,746 @@
+	// result occurs. This seems fairly proxy agnostic.
+	NumRetries uint32 `json:",omitempty"`
+
+	// RetryOnConnectFailure allows for connection failure errors to trigger a
+	// retry. This should be expressible in other proxies as it's just a layer
+	// 4 failure bubbling up to layer 7.
+	RetryOnConnectFailure bool `json:",omitempty"`
+
+	// RetryOnStatusCodes is a flat list of http response status codes that are
+	// eligible for retry. This again should be feasible in any sane proxy.
+	RetryOnStatusCodes []uint32 `json:",omitempty"`
+}
+
+// ServiceSplitterConfigEntry defines how incoming requests are split across
+// different subsets of a single service (like during staged canary rollouts),
+// or perhaps across different services (like during a v2 rewrite or other type
+// of codebase migration).
+//
+// This config entry represents the next hop of the discovery chain after
+// routing. If no splitter config is defined the chain assumes 100% of traffic
+// goes to the default service and discovery continues on to the resolution
+// hop.
+//
+// Splitter configs are recursively collected while walking the discovery
+// chain.
+//
+// Splitter config entries will be restricted to only services that define
+// their protocol as http-based (in centralized configuration).
+type ServiceSplitterConfigEntry struct {
+	Kind string
+	Name string
+
+	// Splits is the configurations for the details of the traffic splitting.
+	//
+	// The sum of weights across all splits must add up to 100.
+	//
+	// If the split is within epsilon of 100 then the remainder is attributed
+	// to the FIRST split.
+	Splits []ServiceSplit
+
+	RaftIndex
+}
+
+func (e *ServiceSplitterConfigEntry) GetKind() string {
+	return ServiceSplitter
+}
+
+func (e *ServiceSplitterConfigEntry) GetName() string {
+	if e == nil {
+		return ""
+	}
+
+	return e.Name
+}
+
+func (e *ServiceSplitterConfigEntry) Normalize() error {
+	if e == nil {
+		return fmt.Errorf("config entry is nil")
+	}
+
+	e.Kind = ServiceSplitter
+
+	// This slightly massages inputs to enforce that the smallest representable
+	// weight is 1/10000 or .01%
+
+	if len(e.Splits) > 0 {
+		sumScaled := 0
+		for i, split := range e.Splits {
+			weightScaled := scaleWeight(split.Weight)
+			e.Splits[i].Weight = float32(float32(weightScaled) / 100.0)
+
+			sumScaled += weightScaled
+		}
+	}
+
+	return nil
+}
+
+func (e *ServiceSplitterConfigEntry) Validate() error {
+	if e.Name == "" {
+		return fmt.Errorf("Name is required")
+	}
+
+	if len(e.Splits) == 0 {
+		return fmt.Errorf("no splits configured")
+	}
+
+	const maxScaledWeight = 100 * 100
+
+	copyAsKey := func(s ServiceSplit) ServiceSplit {
+		s.Weight = 0
+		return s
+	}
+
+	// Make sure we didn't refer to the same thing twice.
+	found := make(map[ServiceSplit]struct{})
+	for _, split := range e.Splits {
+		splitKey := copyAsKey(split)
+		if splitKey.Service == "" {
+			splitKey.Service = e.Name
+		}
+		if _, ok := found[splitKey]; ok {
```

Comment:
Does using a value of struct as a key behave as we like here? Does it implicitly use pointer comparison or is there actual deep-equality being checked somehow? I'm not sure I've seen this in Go before.

Reply:
* https://blog.golang.org/go-maps-in-action#TOC_5.
* https://golang.org/ref/spec#Map_types
* https://play.golang.org/p/EI25fdD0Fcj

So long as the struct is comparable (defined above) it can be a key. If you happen to do something with pointers (which I would not advise due to sanity) only the addresses are compared:
> Pointer values are comparable. Two pointer values are equal if they point to the same variable or if both have value nil. Pointers to distinct zero-size variables may or may not be equal.

In the examples used in this PR I intended to use only non-pointer primitives in the key structs where simple `==` works as you would expect.